### PR TITLE
wrote ITPParser

### PR DESCRIPTION
**Pull request by lilyminium**
I'd like to read some ITP files.

Changes made in this Pull Request:
 - Added parser for ITP files

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

File: package/CHANGELOG [hunk 1]
```diff
@@ -27,6 +27,7 @@ Fixes
 Enhancements
   * Expanded selection wildcards to the start and middle of strings (Issue #2370)
   * Added type checking and conversion to Connection TopologyAttrs (Issue #2373)
+  * Added ITPParser
```
Comment (richardjgowers) [inline package/CHANGELOG]: Expand this Gromacs ITP topology or whatever

File: package/MDAnalysis/topology/ITPParser.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    - resids
+    - resnames
+    - segids
+    - moltypes
+    - bonds
+    - angles
+    - dihedrals
+    - impropers
+
+    .. _ITP: http://www.charmm.org/documentation/c35b1/struct.html
```
Comment (richardjgowers) [inline package/MDAnalysis/topology/ITPParser.py]: Charmm 🤨

File: package/MDAnalysis/topology/ITPParser.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        impropers = []
+
+        segid = 'SYSTEM'
+        molnum = 1
+
+        section = None
+
+        # Open and check itp validity
+        with openany(self.filename) as itpfile:
+            header = re.compile(r'\[\s*(?P<section>\w+)\s*\]')
```
Comment (richardjgowers) [inline package/MDAnalysis/topology/ITPParser.py]: Fwiw I think usually regex is slower than just checking for a left bracket and it looks like this is being called on all lines. I guess itps are probably gonna be small though

File: package/MDAnalysis/topology/ITPParser.py (hunk 2 above)
Comment (lilyminium) [reply package/MDAnalysis/topology/ITPParser.py]: Oops!

File: package/MDAnalysis/topology/ITPParser.py (hunk 3 above)
Comment (lilyminium) [reply package/MDAnalysis/topology/ITPParser.py]: Changed it anyway, and updated CHANGELOG

**Comment by lilyminium** [comment]
I added bond/angle/dihedral/improper types. They are strings to match what appears to be the behaviour of #763 and #774.

File: testsuite/MDAnalysisTests/datafiles.py [hunk 4]
```diff
@@ -473,5 +476,12 @@
 PDB_rama = resource_filename(__name__, 'data/19hc.pdb.gz')
 PDB_janin = resource_filename(__name__, 'data/1a28.pdb.gz')
 
+ITP_ala = resource_filename(__name__, 'data/gromacs_ala10.itp')
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/datafiles.py]: this is a sneaky gotcha, but new data files need to get added to the testsuite setup.py so they get bundled up into the tarballs of the release. There’ll be a bunch of glob patterns but I don’t think the new itps will get caught

File: testsuite/MDAnalysisTests/datafiles.py [hunk 5]
```diff
@@ -473,5 +476,12 @@
...
+XTC_ala = resource_filename(__name__, 'data/gromacs_ala10.xtc')
+
+ITP_atb = resource_filename(__name__, 'data/atb_hema.itp')
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/datafiles.py]: Also these files are unused so either use them or lose them. You can also compress the files and MDA should just work still

File: package/MDAnalysis/topology/ITPParser.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+   :members:
+   :inherited-members:
+
+"""
+from __future__ import absolute_import, division
+from six.moves import range
+
+import logging
+import functools
+import re
```
Comment (richardjgowers) [inline package/MDAnalysis/topology/ITPParser.py]: Unused imports?

**Comment by lilyminium** [comment]
@richardjgowers is MDAnalysis interested in topology writers? (I would also like to write the ITPs out). All the Writers seem to be associated with coordinates.

File: package/doc/sphinx/source/documentation_pages/topology_modules.rst [hunk 7]
```diff
@@ -31,6 +31,7 @@ topology file format in the *topology_format* keyword argument to
    topology/GMSParser
    topology/GROParser
    topology/HoomdXMLParser
+   topology/ITPParser
```
Comment (richardjgowers) [inline package/doc/sphinx/source/documentation_pages/topology_modules.rst]: You'll need to add a .rst stub for this new module

**Comment by richardjgowers** [comment]
@lilyminium yeah I think topology writers could make sense... Though they often contain a superset of information than a Universe (ie force field information)

**Comment by orbeckst** [comment]
Topology writers are a thorny issue. It's really hard to get them right – do you believe that the written ITPs will be good enough to run in Gromacs?

There are ParmEd and intermol, which contain an awful lot of knowledge about force fields and topologies. I'd rather leave topology writing/hacking to them instead of having to deal with the problems.

**Comment by orbeckst** [comment]
I'd rather look into the question how MDAnalysis can call ParmEd to write a topology.

**Comment by lilyminium** [comment]
Yes... last I checked (last year) ParmEd refused to read GROMACS topology files with G96 type information (e.g. quartic bond potentials), which also happen to be the most common ones for native GROMOS force fields. (I haven't looked at InterMol yet). Converting quartic potentials to harmonics for another force field does sound like a terrible idea, but being able to load in a topology, select certain atoms, and write out the applicable bonds/angles/dihedrals/impropers to the same format is quite useful. Or even renumbering atoms, etc. 

Faithfully reproducing ITP files in MDAnalysis would be a pain because you can have multiple entries for the same bond, etc. in an ITP file that work additively. (That being said I've found the ITP file spec reasonably forgiving regarding whitespace and so on). 

Using ParmEd from MDAnalysis would be interesting.

**Pull request merged**